### ISSUE TITLE
Installing from git (or a git snapshot) results in package version being 0

### ISSUE THREAD

**Issue by mgorny** [samuelcolvin/dirty-equals#45]
I've just noticed that when using git snapshots or the git repository itself, dirty-equals ends up being installed as version "0", e.g.:

```
$ pip install git+https://github.com/samuelcolvin/dirty-equals
Collecting git+https://github.com/samuelcolvin/dirty-equals
  Cloning https://github.com/samuelcolvin/dirty-equals to /tmp/pip-req-build-a0xri30s
  Running command git clone --filter=blob:none --quiet https://github.com/samuelcolvin/dirty-equals /tmp/pip-req-build-a0xri30s
  Resolved https://github.com/samuelcolvin/dirty-equals to commit 593bcccf738ab8b724d7cb860881d74344171f5f
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: pytz>=2021.3 in ./.venv/lib/python3.11/site-packages (from dirty-equals==0) (2022.1)
Building wheels for collected packages: dirty-equals
  Building wheel for dirty-equals (pyproject.toml) ... done
  Created wheel for dirty-equals: filename=dirty_equals-0-py3-none-any.whl size=23380 sha256=96b6eecea7fe5c51bc14f23929d3f6146e22344aff7c82749c96fb92c414218e
  Stored in directory: /tmp/pip-ephem-wheel-cache-qjkkipja/wheels/b9/a6/72/32198aa9ff0bfe14230160fbf92821af4a22179c61d55c7070
Successfully built dirty-equals
Installing collected packages: dirty-equals
Successfully installed dirty-equals-0
```

**Comment by mgorny** [comment]
I've tried adding `poetry-version-plugin` into `build-system.requires` but that doesn't seem to help (beside pulling it tons of build-time dependencies).

**Comment by samuelcolvin** [comment]
Ye, we set the version in CI based on the tag:

https://github.com/samuelcolvin/dirty-equals/blob/593bcccf738ab8b724d7cb860881d74344171f5f/.github/workflows/ci.yml#L129

Which in turn calls https://gist.github.com/samuelcolvin/e49096dc93396be60d2e972e8799410d I don't really want to change this as I use similar logic for most projects and it avoids having to set the version in code before creating a release.

**Comment by samuelcolvin** [comment]
Happy to change the default value if there's something more sensible, but I don't see another solution.

**Comment by mgorny** [comment]
Well, that sucks for us packagers but I guess it's not the end of the world.

I wish there was a way for this to either fail or at least emit verbose warning when the value's not updated. I've only noticed this because `pip list --outdated` listed the version as `0`. In general, potential version problems like that are very hard to notice until something suddenly explodes for end users.

**Comment by samuelcolvin** [comment]
humm, ye I feel your pain.

I guess we could push to the repr on release to set the version in code - that would be neat generally. Though I guess it might not help if you use the version from the version tag?

**Comment by mgorny** [comment]
As a minor request, could you please use the same version string both in `__init__.py` and `pyproject.toml`, to make the substitution easier for us?

**Comment by mgorny** [comment]
> I guess we could push to the repr on release to set the version in code - that would be neat generally. Though I guess it might not help if you use the version from the version tag?

Took me a while to figure it out but I guess you meant "push to the repo"? Yeah, that wouldn't help if you pushed after tagging. I guess it would be somewhat ugly but you could make the CI pipeline rewrite the tag though.

**Comment by samuelcolvin** [comment]
@mgorny worth noting you're going  to have even more problems with watchfiles where we do something similar, but have logic to convert a python style version to rust before editing `Cargo.toml`, then more logic to convert a rust style version back to python at runtime, see https://github.com/samuelcolvin/watchfiles/pull/159.

And also with pydantic-core where I will do something similar.

**Comment by samuelcolvin** [comment]
Just thinking about this, would it help if we edited `pyproject.toml` at the same time as `__init__.py` so at least the source and wheels on pypi were right?

**Comment by mgorny** [comment]
> @mgorny worth noting you're going to have even more problems with watchfiles where we do something similar, but have logic to convert a python style version to rust before editing `Cargo.toml`, then more logic to convert a rust style version back to python at runtime, see [samuelcolvin/watchfiles#159](https://github.com/samuelcolvin/watchfiles/pull/159).

Well, I'm just using `sed` to put the correct version into `Cargo.toml` there.

**Comment by samuelcolvin** [comment]
Should work fine except on pre-releases.

**Comment by samuelcolvin** [comment]
This should be fully fixed by #49 which sets the version in git properly.
